packages/server/graphql/private/mutations/removeAuthIdentity.ts

```diff
@@ -0,0 +1,80 @@
+import base64url from 'base64url'
+import bcrypt from 'bcrypt'
+import crypto from 'crypto'
+import {AuthIdentityTypeEnum, Security} from 'parabol-client/types/constEnums'
+import {r} from 'rethinkdb-ts'
+import util from 'util'
+import AuthIdentityLocal from '../../../database/types/AuthIdentityLocal'
+import PasswordResetRequest from '../../../database/types/PasswordResetRequest'
+import getMailManager from '../../../email/getMailManager'
+import resetPasswordEmailCreator from '../../../email/resetPasswordEmailCreator'
+import generateRandomString from '../../../generateRandomString'
+import getUsersByDomain from '../../../postgres/queries/getUsersByDomain'
+import updateUser from '../../../postgres/queries/updateUser'
+import {MutationResolvers} from '../../private/resolverTypes'
+
+const randomBytes = util.promisify(crypto.randomBytes)
+
+const removeAuthIdentity: MutationResolvers['removeAuthIdentity'] = async (
+  _source,
+  {domain, identityType, addLocal},
+  {ip}
+) => {
+  // VALIDATION
+  const users = await getUsersByDomain(domain)
+  if (!users.length) {
+    return {error: {message: 'No user emails match that domain'}}
+  }
+
+  // RESOLUTION
+  const updateUsersPromises = users.map(async (user) => {
+    const {id: userId, identities} = user
+    const filteredIdentities = identities.filter((identity) => identity.type !== identityType)
+    if (!filteredIdentities.length || addLocal) {
+      const identityId = `${userId}:${AuthIdentityTypeEnum.LOCAL}`
+      const dummyPassword = generateRandomString(Security.SALT_ROUNDS)
+      const dummyHashedPassword = await bcrypt.hash(dummyPassword, Security.SALT_ROUNDS)
+      const newIdentity = new AuthIdentityLocal({
+        hashedPassword: dummyHashedPassword,
+        id: identityId,
+        isEmailVerified: false
+      })
+      const newIdentities = [...filteredIdentities, newIdentity]
+      return updateUser({identities: newIdentities}, userId)
+    } else {
+      return updateUser({identities: filteredIdentities}, userId)
+    }
+  })
+  await Promise.all(updateUsersPromises)
+
+  // send forgot password email
+  users.forEach(async ({email}) => {
```

Comment:
If a domain update affects hundreds of users and mailgun is being slow, could this mutation timeout? We want to invalidate all the tokens, and then insert the new password request, and then send the email, so I think we need to `await` each time.

Reply:
-1 `forEach` isn't equipped to handle async functions.
consider the following 

```ts
[1, 2, 3].forEach(async () => {
    await new Promise((resolve) =>
      setTimeout(() => {
        console.log('resolving')
        resolve()
      }, 1000)
    )
  })
console.log('do i print before or after?')
```
what you want to do is call `Promise.all(users.map(...)`

here's what I'd do...
1. instead of updating the user in `updateUsersPromises`, have that function return an array of users with updated identities.
2. extract out the chunk that was copied from `emailPasswordReset`
3. call `Promise.all(usersWithUpdatedIdentities.map((user) => processEmailPasswordReset(ip, user.identities, user.id...))`